### ISSUE TITLE
Generic KeySerializer / Deserializer for Serializable

### ISSUE THREAD

**Issue by freddyboucher** [nmorel/gwt-jackson#124]
Hi,
I'm creating a new [gwt-jackson-objectify](https://github.com/freddyboucher/gwt-jackson/tree/objectify)

So far everything is working fine now!

But in addition of registering this new objectify extension in my project with `<inherits name="com.github.nmorel.gwtjackson.objectify.GwtJacksonObjectify" />`

I also have to declare `<extend-configuration-property name="gwtjackson.configuration.extension" value="com.learnkeeper.client.util.EntitiesJacksonConfiguration" />`

with the following **EntitiesJacksonConfiguration**:

```java
import com.github.nmorel.gwtjackson.client.AbstractConfiguration;
import com.github.nmorel.gwtjackson.client.JsonDeserializationContext;
import com.github.nmorel.gwtjackson.client.JsonSerializationContext;
import com.github.nmorel.gwtjackson.client.ObjectMapper;
import com.github.nmorel.gwtjackson.client.deser.map.key.KeyDeserializer;
import com.github.nmorel.gwtjackson.client.ser.map.key.KeySerializer;
import com.google.gwt.core.client.GWT;

import com.learnkeeper.shared.entities.Record;

public class EntitiesJacksonConfiguration extends AbstractConfiguration {

  @Override
  protected void configure() {
    key(Record.class).serializer(RecordKeySerializer.class).deserializer(RecordKeyDeserializer.class);
  }

  public interface RecordObjectMapper extends ObjectMapper<Record> {
    RecordObjectMapper INSTANCE = GWT.create(RecordObjectMapper.class);
  }

  public static class RecordKeySerializer extends KeySerializer<Record> {

    private static final RecordKeySerializer INSTANCE = new RecordKeySerializer();

    private RecordKeySerializer() { }

    public static RecordKeySerializer getInstance() {
      return INSTANCE;
    }

    @Override
    protected String doSerialize(Record value, JsonSerializationContext ctx) {
      return RecordObjectMapper.INSTANCE.write(value, ctx);
    }
  }

  public static class RecordKeyDeserializer extends KeyDeserializer<Record> {

    private static final RecordKeyDeserializer INSTANCE = new RecordKeyDeserializer();

    private RecordKeyDeserializer() { }

    public static RecordKeyDeserializer getInstance() {
      return INSTANCE;
    }

    @Override
    protected Record doDeserialize(String key, JsonDeserializationContext ctx) {
      return RecordObjectMapper.INSTANCE.read(key, ctx);
    }
  }
}
```

```java
import java.io.Serializable;

import com.googlecode.objectify.annotation.Entity;

@Entity
public class Record implements Serializable {

}
```

I have to do that because I want to use my **Record** object as Map key.

What I would like is to include the ability of using `@Entity` or  `Serializable` as Map key directly in the objectify extension.

This is my attempt to do it: https://github.com/freddyboucher/gwt-jackson/commit/b1371f5bf6e2a46f4aa41d634ac99aa6aca79773

Is it possible to achieve it or should I keep my current implementation and define for each of my `@Entity` a custom key Serializer / Deserializer in `EntitiesJacksonConfiguration`?

Thanks so much

**Comment by freddyboucher** [comment]
Since I started this ticket, I slightly changed the approach!

# 1. Server side

I have now a working **server** side implementation:

If I want to use any **Bean** as Map key, I just have to implement a new **generic** interface [KeySerializable\<T>](https://github.com/freddyboucher/gwt-jackson/blob/fcd5eeb965c24974587c0257b026254aec2b7c07/extensions/objectify/src/main/java/com/github/nmorel/gwtjackson/objectify/shared/KeySerializable.java).

Related code: https://github.com/freddyboucher/gwt-jackson/commit/fcd5eeb965c24974587c0257b026254aec2b7c07#diff-154b377ff849dd5baf3d79b0bd4cc77a
```java
@Entity
public class Bean implements KeySerializable<Bean> {
}
```

Then where I instantiate my ObjectMapper, I have to register the real Bean as key.
Related code: https://github.com/freddyboucher/gwt-jackson/commit/fcd5eeb965c24974587c0257b026254aec2b7c07
```java
        objectMapper = new ObjectMapper();
        ObjectifyJacksonModule module = new ObjectifyJacksonModule();
        module.addKeySerializer( Bean.class, new KeySerializableSerializer() );
        module.addKeyDeserializer( Bean.class, new KeySerializableDeserializer() );
        objectMapper.registerModule( module );
```

# 2. Client side

Not working yet, it probably requires a little change in the **gwt-jackson** core library.

I have to create a new **AbstractConfiguration** class to register my **Bean**.
Related code: https://github.com/freddyboucher/gwt-jackson/commit/fcd5eeb965c24974587c0257b026254aec2b7c07#diff-d6b20c7fd22cd4340fe17c961fcfa7b9
```java
public class BeanObjectifyConfiguration extends AbstractConfiguration {

    @Override
    protected void configure() {
        key( Bean.class ).serializer( KeySerializableSerializer.class ).deserializer( KeySerializableDeserializer.class );
    }
}
```

And then off course, I have to declare this new **AbstractConfiguration** in my project module.gwt.xml
Related code: https://github.com/freddyboucher/gwt-jackson/commit/fcd5eeb965c24974587c0257b026254aec2b7c07#diff-2df6292d6b5df45bca7bd484a8afda81
```xml
<extend-configuration-property name="gwtjackson.configuration.extension"
                                 value="com.github.nmorel.gwtjackson.objectify.client.BeanObjectifyConfiguration" />
```

It's not working because **gwt-jackson** does't map properly the **key** (my Bean object) and the custom **KeySerializableSerializer** / **KeySerializableDeserializer** because it's based on an [KeySerializable](https://github.com/freddyboucher/gwt-jackson/blob/fcd5eeb965c24974587c0257b026254aec2b7c07/extensions/objectify/src/main/java/com/github/nmorel/gwtjackson/objectify/shared/KeySerializable.java) **interface** rather than a **class**.
>Rebinding com.github.nmorel.gwtjackson.objectify.client.KeySerializableTestCase.TestObjectMapper
   Invoking generator com.github.nmorel.gwtjackson.rebind.ObjectMapperGenerator
      [WARN] Cannot instantiate the custom key serializer/deserializer com.github.nmorel.gwtjackson.objectify.client.KeySerializableSerializer. It will be ignored
      [WARN] Cannot instantiate the custom key serializer/deserializer com.github.nmorel.gwtjackson.objectify.client.KeySerializableDeserializer. It will be ignored
      [WARN] Type 'com.github.nmorel.gwtjackson.objectify.shared.Bean' is not supported as map's key
      [ERROR] Cannot generate mapper due to previous errors : Type 'com.github.nmorel.gwtjackson.objectify.shared.Bean' is not supported as map's key
[ERROR] Deferred binding failed for 'com.github.nmorel.gwtjackson.objectify.client.KeySerializableTestCase.TestObjectMapper'; expect subsequent failures

@nmorel 
Do you think it's feasible? That's the last piece before I can submit this new extension.

Thanks so much for your help.

**Comment by dnouls** [comment]
Any progress on supporting any key type in gwt jackson ? I need this as well in my project.

**Comment by freddyboucher** [comment]
@dnouls 
Yes it's already done by @nmorel  here https://github.com/nmorel/gwt-jackson/pull/128/commits/d9aabdb94e51137b21db688db1fe9c39187b8a07
It's part of another PR but I can easily submit a PR specifically to resolve this issue to unblock you.
I will do that later today.

**Comment by freddyboucher** [comment]
@dnouls 
I've created the PR https://github.com/nmorel/gwt-jackson/pull/135
